csrc/fusion_segmenter.cpp

```diff
@@ -3848,14 +3850,54 @@ void SegmentCandidateFinder::findSegments() {
 
   segmented_fusion_->validateIfDebug();
 
+  // Resolve all the input expressions needed in each group
+  for (auto [forwarded_input, aux_group] : input2group_) {
+    if (forwarded_input->isFusionInput()) {
+      // Nothing to resolve.
+      continue;
+    }
+
+    if (forwarded_input->isScalar()) {
+      // Will be resolved after this loop.
+      continue;
+    }
+
+    resolveInputGroup(aux_group, forwarded_input);
+    // aux_group will be removed from segmented_fusion_ by
+    // cleanupForwardedInputs.
+  }
+
+  // Un-forward scalar inputs unconditionally. The loop above doesn't resolve
```

Comment:
Is this necessary? After this part, there's a call to `finalize`, which calles `resolveScalarsInGroup`. All necessary scalars included forwarded ones are resolved there.

Reply:
That's what I initially thought but FusionExpand_CUDA (and some others) failed without this. The failure was around https://github.com/NVIDIA/Fuser/blob/925acc63e5870718ab83de911ec7f85d6126f729/csrc/kernel_cache.cpp#L400-L404. I saw a scalar group input was never ready because it wasn't a fusion input. If you believe resolveScalarsInGroup should forward scalar inputs, then it must be a bug there.

Reply:
Ahh, makes sense. It was previously done in `resolveInputsInGroup` together with tensor inputs.